quantumnetworks/analysis/error.py

```diff
@@ -0,0 +1,120 @@
+from typing import Any, Dict
+from abc import abstractmethod, ABCMeta
+
+from quantumnetworks.systems.multimode import MultiModeSystem
+from quantumnetworks.utils.visualization import plot_full_evolution
+from tqdm import tqdm
+
+
+class SystemError(metaclass=ABCMeta):
+    def __init__(self, system: SystemSolver, params_error: Dict[str, Any]) -> None:
+        self.system = system
+        self.params_error = params_error
+        self.solves = None
+        self.reset_solves()
+
+    def reset_solves(self):
+        self.solves = {"original": None, "with_error": [], "std": None}
+
+    @abstractmethod
+    def calculate_error(self, method: str, *args, parse_output=lambda X: X, **kwargs):
+        pass
+
+    def run(
+        self, method: str, *args, parse_output=lambda X: X, num_samples=11, **kwargs
+    ):
+        self.solves["original"] = parse_output(
+            getattr(self.system, method)(*args, **kwargs)
+        )
+        self.calculate_error(
+            method, *args, parse_output=parse_output, num_samples=num_samples, **kwargs
+        )
+        solves_with_error = np.array(self.solves["with_error"])
+        self.solves["std"] = np.std(solves_with_error, axis=0)
+
+    def plot(self, ts, **kwargs):
+        fig, ax = plot_full_evolution(
+            self.solves["original"],
+            ts,
+            xs_min=(self.solves["original"] - self.solves["std"]),
+            xs_max=(self.solves["original"] + self.solves["std"]),
+            **kwargs
+        )
+        ax.legend()
+        fig.tight_layout()
+        return fig, ax
+
+
+class MultiModeError(SystemError):
+    def calculate_error(
+        self, method: str, *args, num_samples=11, parse_output=lambda X: X, **kwargs
+    ):
+        # =====================
+        def nonnegative(a):
+            a[a < 0] = 0
+            return a
+
+        def couplings_list2dict(cps):
+            # TODO: remove when couplings are represented by dict
+            return {(min(row[0], row[1]), max(row[0], row[1])): row[2] for row in cps}
+
+        def couplings_dict2list(cps):
+            return np.array([[key[0], key[1], val] for key, val in cps.items()])
+
+        # =====================
+        # error
+        omegas_error = np.array(self.params_error["omegas"])
+        kappas_error = np.array(self.params_error["kappas"])
+        gammas_error = np.array(self.params_error["gammas"])
+        kerrs_error = np.array(self.params_error["kerrs"])
+        couplings_error = couplings_list2dict(np.array(self.params_error["couplings"]))
+
+        params_original = self.system.params.copy()
+        couplings_original_dict = couplings_list2dict(params_original["couplings"])
+
+        params = self.system.params.copy()
+
+        num_modes = params_original["num_modes"]
+        num_variables = 4 * num_modes + len(couplings_original_dict)
+        scalings = np.random.normal(size=num_samples * num_variables).reshape(
+            num_samples, num_variables
+        )
```

Comment:
Minor, but you might be able to specify the shape directly in size, without needing `np.reshape`. 
E.g. `size=(num_samples, num_variables)`

Reply:
Just to confirm, but this is assuming the error on all omegas or kappas, etc. is the same? That is `omegas_error` is assumed to be the same for `omega_a`, `omega_b`, etc? Might be good to mention that explicitly in a docstring since in principle one might have different uncertainties on those (e.g. if `omega_a` is a microwave resonator and `omega_b` is a Fabry-Perot optical cavity, the experiments one would do to measure the frequencies are different and so might have different uncertainties). 

That said, keeping `omegas_error` the same for all is totally fine here (i.e. not worth the time to have to explicitly set all uncertainties manually). So maybe just mention it as a specific choice and leave it at that

Reply:
No, `omegas_error` is a vector and we're doing element-wise multiplication when we do `omegas_error * scaling_vec[:num_modes]`!

Check the notebook for detailed use!